File: java/src/org/sanity/consoleForum/commands/DislikeCommand.java
package org.sanity.consoleForum.commands;

import org.sanity.consoleForum.commands.Attributes.*;
import org.sanity.consoleForum.common.Constants;
import org.sanity.consoleForum.core.Principal;
import org.sanity.consoleForum.database.EntityManager;
import org.sanity.consoleForum.io.InputReader;
import org.sanity.consoleForum.io.OutputWriter;
import org.sanity.consoleForum.models.Post;
import org.sanity.consoleForum.models.PostRating;
import org.sanity.consoleForum.models.enums.PostRatingChoice;

import java.io.IOException;
import java.util.List;

@UserCommand
@ModeratorCommand
@AdminCommand
@CommandHelper(help = "Dislikes a Post by a given Post Id.")
@CommandExample(example = "Dislike|1")
public class DislikeCommand implements Command {
    private static final String NoSuchPostMessage = "There is no Post with the given Id.";

    private static final String CannotDislikeAgainMessage = "You already disliked this Post.";

    private static final String SuccessDislikeMessage = "Successfully Disliked Post - %s!";

    private InputReader consoleReader;

    private OutputWriter consoleWriter;

    private List<String> parameters;

    private Principal principal;

    private boolean resetsConsole;

    private EntityManager entityManager;

    @Override
    public InputReader getConsoleReader() {
        return this.consoleReader;
    }

    @Override
    public void setConsoleReader(InputReader inputReader) {
        this.consoleReader = inputReader;
    }

    @Override
    public OutputWriter getConsoleWriter() {
        return this.consoleWriter;
    }

    @Override
    public void setConsoleWriter(OutputWriter outputWriter) {
        this.consoleWriter = outputWriter;
    }

    @Override
    public List<String> getParameters() {
        return this.parameters;
    }

    @Override
    public void setParameters(List<String> parameters) {
        this.parameters = parameters;
    }

    @Override
    public Principal getPrincipal() {
        return this.principal;
    }

    @Override
    public void setPrincipal(Principal principal) {
        this.principal = principal;
    }

    @Override
    public boolean getResetsConsole() {
        return this.resetsConsole;
    }

    @Override
    public void setResetsConsole(boolean resetsConsole) {
        this.resetsConsole = resetsConsole;
    }

    @Override
    public EntityManager getEntityManager() {
        return entityManager;
    }

    @Override
    public void setEntityManager(EntityManager entityManager) {
        this.entityManager = entityManager;
    }

    public void execute() throws InterruptedException, IOException {
        int postId = Integer.parseInt(this.getParameters().get(0));

        Post postFromDb = this.getEntityManager().getPosts()
                .stream()
                .filter(post -> post.getId() == postId)
                .findFirst()
                .orElse(null);

        if (postFromDb == null || postFromDb.getIsDeleted())
        {
            this.consoleWriter.write(Constants.CONSOLE_FORUM_OUTPUT_PREFIX);
            this.consoleWriter.errorLine(NoSuchPostMessage);
            Thread.sleep(Constants.NOTIFICATION_DELAY);
        }
        else if (postFromDb.getRatings().stream().anyMatch(rating -> rating.getUser().getId()
                == this.getPrincipal().getUser().getId()))
        {
            PostRating ratingFromDb = postFromDb.getRatings()
                    .stream()
                    .filter(rating -> rating.getUser().getId() == this.getPrincipal().getUser().getId())
                    .findFirst()
                    .orElse(null);

            if (ratingFromDb.getIsNegative())
            {
                this.consoleWriter.write(Constants.CONSOLE_FORUM_OUTPUT_PREFIX);
                this.consoleWriter.errorLine(CannotDislikeAgainMessage);
                Thread.sleep(Constants.NOTIFICATION_DELAY);
            }
            else
            {
                ratingFromDb.toggle();

                this.consoleWriter.write(Constants.CONSOLE_FORUM_OUTPUT_PREFIX);
                this.consoleWriter.successLine(SuccessDislikeMessage, postId);
                Thread.sleep(Constants.NOTIFICATION_DELAY);
            }
        }
            else
        {
            PostRating rating = new PostRating(PostRatingChoice.NEGATIVE) {{
                setUser(getPrincipal().getUser());
                setPost(postFromDb);
            }};

            this.getEntityManager().add(rating);

            this.getPrincipal().getUser().getRatings().add(rating);
            postFromDb.getRatings().add(rating);

            this.consoleWriter.write(Constants.CONSOLE_FORUM_OUTPUT_PREFIX);
            this.consoleWriter.successLine(SuccessDislikeMessage, postId);
            Thread.sleep(Constants.NOTIFICATION_DELAY);
        }
    }
}

File: csharp/src/ConsoleForum.App/Commands/DislikeCommand.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading;
using ConsoleForum.App.Commands.Attributes;
using ConsoleForum.App.Common;
using ConsoleForum.App.Core;
using ConsoleForum.App.Database;
using ConsoleForum.App.IO;
using ConsoleForum.App.Models;
using ConsoleForum.App.Models.Enums;

namespace ConsoleForum.App.Commands
{
    [UserCommand]
    [ModeratorCommand]
    [AdminCommand]
    [CommandHelper("Dislikes a Post by a given Post Id.")]
    [CommandExample("Dislike|1")]
    public class DislikeCommand : ICommand
    {
        private const string NoSuchPostMessage = "There is no Post with the given Id.";

        private const string CannotDislikeAgainMessage = "You already disliked this Post.";

        private const string SuccessDislikeMessage = "Successfully Disliked Post - {0}!";

        public IInputReader ConsoleReader { get; set; }

        public IOutputWriter ConsoleWriter { get; set; }

        public List<string> Parameters { get; set; }

        public IPrincipal Principal { get; set; }

        public bool ResetsConsole { get; set; } = true;

        public void Execute()
        {
            int postId = int.Parse(this.Parameters[0]);

            Post postFromDb = DbContext.Posts.FirstOrDefault(post => post.Id == postId);

            if (postFromDb == null || postFromDb.IsDeleted)
            {
                this.ConsoleWriter.Write(Constants.ConsoleForumOutputPrefix);
                this.ConsoleWriter.ErrorLine(NoSuchPostMessage);
                Thread.Sleep(Constants.NotificationDelay);
            }
            else if (postFromDb.Ratings.Any(rating => rating.User.Id == this.Principal.User.Id))
            {
                PostRating ratingFromDb = postFromDb.Ratings.First(rating => rating.User.Id == this.Principal.User.Id);

                if (ratingFromDb.IsNegative)
                {
                    this.ConsoleWriter.Write(Constants.ConsoleForumOutputPrefix);
                    this.ConsoleWriter.ErrorLine(CannotDislikeAgainMessage);
                    Thread.Sleep(Constants.NotificationDelay);
                }
                else
                {
                    ratingFromDb.Toggle();

                    this.ConsoleWriter.Write(Constants.ConsoleForumOutputPrefix);
                    this.ConsoleWriter.SuccessLine(SuccessDislikeMessage, postId);
                    Thread.Sleep(Constants.NotificationDelay);
                }
            }
            else
            {
                PostRating rating = new PostRating(PostRatingChoice.Negative)
                {
                    User = this.Principal.User,
                    Post = postFromDb
                };

                DbContext.Add(rating);

                this.Principal.User.Ratings.Add(rating);
                postFromDb.Ratings.Add(rating);

                this.ConsoleWriter.Write(Constants.ConsoleForumOutputPrefix);
                this.ConsoleWriter.SuccessLine(SuccessDislikeMessage, postId);
                Thread.Sleep(Constants.NotificationDelay);
            }
        }
    }
}
